tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,570 @@
+type ErrInputOutputDataTypeMismatch struct {
+	outDt DataType
+	inDt  DataType
+}
+
+func (e *ErrInputOutputDataTypeMismatch) Error() string {
+	return fmt.Sprintf("The output datatype '%s' doesn't correspond with the input data type '%s'",
+		e.outDt, e.inDt)
+}
+
+// castAttrValue Returns an pb.AttrValue that contains the corresponding
+// pb.AttrValue_* according to the type specified. Returns nil if the data type
+// of the provided value can't be allocated on the AttrValue type
+func (gr *Graph) castAttrValue(attrType string, v interface{}) (attrVal *pb.AttrValue) {
+	switch attrType {
+	case "type":
+		if dt, ok := v.(DataType); ok {
+			return &pb.AttrValue{
+				Value: &pb.AttrValue_Type{
+					Type: pb.DataType(dt),
+				},
+			}
+		}
+	case "string":
+		if st, ok := v.(string); ok {
+			return &pb.AttrValue{
+				Value: &pb.AttrValue_S{
+					S: []byte(st),
+				},
+			}
+		}
+	case "tensor":
+		if t, ok := v.(*Tensor); ok {
+			tp := &pb.TensorProto{
+				Dtype:         t.Dtype,
+				TensorShape:   t.TensorShape,
+				TensorContent: t.TensorContent,
+			}
+			switch t.DataType() {
+			case DtFloat:
+				tp.FloatVal, _ = t.AsFloat32()
+			case DtDouble:
+				tp.DoubleVal, _ = t.AsFloat64()
+			case DtInt8, DtInt16, DtInt32, DtUint8:
+				tp.IntVal, _ = t.AsInt32()
+			case DtInt64:
+				tp.Int64Val, _ = t.AsInt64()
+			case DtBool:
+				tp.BoolVal, _ = t.AsBool()
+			case DtString:
+				tp.StringVal, _ = t.AsStr()
+			default:
+				return
+			}
+
+			return &pb.AttrValue{
+				Value: &pb.AttrValue_Tensor{
+					Tensor: tp,
+				},
+			}
+		}
+	case "func":
+		if f, ok := v.(*pb.NameAttrList); ok {
+			return &pb.AttrValue{
+				Value: &pb.AttrValue_Func{
+					Func: f,
+				},
+			}
+		}
+	case "int":
+		if i, ok := v.(int64); ok {
+			return &pb.AttrValue{
+				Value: &pb.AttrValue_I{
+					I: i,
+				},
+			}
+		}
+	case "bool":
+		if b, ok := v.(bool); ok {
+			return &pb.AttrValue{
+				Value: &pb.AttrValue_B{
+					B: b,
+				},
+			}
+		}
+	case "float":
+		if f, ok := v.(float32); ok {
+			return &pb.AttrValue{
+				Value: &pb.AttrValue_F{
+					F: f,
+				},
+			}
+		}
+	case "shape":
+		if s, ok := v.(*pb.TensorShapeProto); ok {
+			return &pb.AttrValue{
+				Value: &pb.AttrValue_Shape{
+					Shape: s,
+				},
+			}
+		}
+	case "list(type)", "list(int)", "list(shape)", "list(float)":
+		if lv, ok := v.(*pb.AttrValue_ListValue); ok {
+			return &pb.AttrValue{
+				Value: &pb.AttrValue_List{
+					List: lv,
+				},
+			}
+		}
+	}
+
+	return nil
+}
+
+// Constant Creates a tensor that is added as a constant to the Graph with the
+// specified name.
+func (gr *Graph) Constant(name string, data interface{}) (op *GraphNode, err error) {
+	ts, err := NewTensor(data)
+	if err != nil {
+		return
+	}
+	gr.constants[name] = ts
+
+	return gr.Op("Const", name, nil, "", map[string]interface{}{
+		"dtype": ts.DataType(),
+		"value": ts,
+	})
+}
+
+// matchTypes Matches all the input/output parameters with their corresponding
```

Comment:
s/Matches/matches/
s/deducting/deducing/  (check for this globally -- to "deduct" is to remove from something;  to "deduce" is to infer or figure it out).
Period in "parameters, this" -> parameters.  This
s/in case of the matching/if the matching/
s/mas have/must have/